### PR TITLE
[Storage] Fix #27590: `az storage fs directory download`: Check user sytem PATH for azcopy and use CLI config directory for new install

### LINKED ISSUE #27590: az storage fs directory download replaces local azcopy binary with old version
### Describe the bug

When running `az storage fs directory download` command to download a file from an Azure storage account with ADLS Gen2 enabled, the command seems to download azcopy version 10.13.0, overwriting any local version of the binary you have. 

### Related command

`az storage fs directory download --recursive`

### Errors

No Errors

### Issue script & Debug output

Start version
```
$ azcopy --version
azcopy version 10.20.1
INFO: azcopy 10.20.1: A newer version 10.21.0 is available to download
```

Download command
```
$ az storage fs directory download -f snaketest --account-name eacriprodpipelines -d snaketest --recursive
This command is in preview and under development. Reference and support levels: https://aka.ms/CLI_refstatus

There are no credentials provided in your command and environment, we will query for account key for your storage account.
It is recommended to provide --connection-string, --account-key or --sas-token in your command as credentials.

In addition, setting the corresponding environment variables can avoid inputting credentials in your command. Please use --help to get more information about environment variable usage.
Azcopy command: ['/home/********/bin/azcopy', 'copy', 'https://*********.dfs.core.windows.net/snaketest***, 'snaketest', '--recursive']
INFO: Scanning...
INFO: Any empty folders will be processed, because source and destination both support folders
INFO: azcopy: A newer version 10.21.0 is available to download


Job ced81976-7ab6-e249-7ac4-9f237a51ef72 has started
Log file is located at: /home/*********/.azcopy/ced81976-7ab6-e249-7ac4-9f237a51ef72.log

0.0 %, 0 Done, 0 Failed, 20 Pending, 0 Skipped, 20 Total,


Job ced81976-7ab6-e249-7ac4-9f237a51ef72 summary
Elapsed Time (Minutes): 0.0334
Number of File Transfers: 16
Number of Folder Property Transfers: 4
Total Number of Transfers: 20
Number of Transfers Completed: 20
Number of Transfers Failed: 0
Number of Transfers Skipped: 0
TotalBytesTransferred: 18095520
```

```
$ azcopy --version
azcopy version 10.13.0
INFO: azcopy: A newer version 10.21.0 is available to download
```

The storage command seems to have overwritten by binary with an older one!

### Expected behavior

The azure storage fs directory command doesn't downgrade my azcopy version!

### Environment Summary

```
az --version
azure-cli                         2.53.0

core                              2.53.0
telemetry                          1.1.0

Extensions:
interactive                        0.5.3
storage-preview                  1.0.0b1

Dependencies:
msal                            1.24.0b2
azure-mgmt-resource             23.1.0b2

Python location '/usr/bin/python3.9'
Extensions directory '/home/******/.azure/cliextensions'

Python (Linux) 3.9.16 (main, May 31 2023, 12:21:58)
[GCC 8.5.0 20210514 (Red Hat 8.5.0-18)]

Legal docs and information: aka.ms/AzureCliLegal
```

Your CLI is up-to-date.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by calvinhzy**
**Related command**

**Description**

Fix #27590 Check user sytem PATH for azcopy and use it if the version is newer than 10.13.0. If not found or has old version, use CLI config directory for azcopy

**Testing Guide**

**History Notes**

[Storage] Fix #27590: `az storage fs directory download`: Check user sytem PATH for azcopy and use CLI config directory for new install

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Storage

**Comment by jakevc** [comment]
Some more thoughts here.

The desired behavior is for az CLI to use azcopy binary for some sub commands. The issue was that the AZ CLI command would download azcopy binary to an install location that is common for users to install software `~/bin`, replacing the binary that may already exist on a users machine.  

The proposed solution here is to use the azcopy binary that already exists if one can be found.. but that might not be ideal either because az CLI might depend on features of a specific azcopy version (right now it's pinned at 10.13, but the latest version is 10.21.0), causing more people to file issues about the az storage CLI....

A more stable solution that doesn't break the users local version of azcopy, or potentially break az cli commands by using an incompatible version of azcopy, would be to have a more unique default install location for the az cli's pinned version of azcopy. 

For example, suppose I have a local version of azcopy: `~/bin/azcopy`, who's --version is 10.21.0. Then I run az storage command and that installs the *compatible* version of azcopy in it's own unique install location, say `~/.azure/cliextensions/storage-preview/bin/azcopy`. Then az cli can have it's own pinned version of azcopy, and I can have my own pinned version of azcopy, and everyone is happy.

**Comment by calvinhzy** [comment]
@jakevc Having a pinned version in another location is probably a better solution than what I was trying to do here. Will keep the 10.13.0 as the version.

**Comment by evelyn-ys** [comment]
I do think what jake said is a good suggestion. Can we detect whether there's existing `azcopy` installed before CLI downloads and installs specific azcopy? There's no necessary for CLI to pin azcopy's version. Ideally CLI can smoothly call any version installed by customer as CLI just pass through arguments to azcopy

**Comment by evelyn-ys** [comment]
BTW, resource module installs bicep under our config directory, shall we follow the same path?
https://github.com/Azure/azure-cli/blob/3209b9590793c7670cb8a2f286afe5dfd2b3eab5/src/azure-cli/azure/cli/command_modules/resource/_bicep.py#L41-L42

File: src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py [hunk 1]
```diff
@@ -30,10 +31,26 @@ class AzCopy:
...
+                args = ["azcopy", "--version"]
+                out_bytes = subprocess.check_output(args)
+                out_text = out_bytes.decode('utf-8')
+                version = re.findall(r"azcopy version (.+?)\n", out_text)[0]
+                if parse_version(version) >= parse_version(AZCOPY_VERSION):
+                    self.executable = "azcopy"
+            except FileNotFoundError:
+                pass
         self.creds = creds
-        if not os.path.isfile(install_location) or self.check_version() != AZCOPY_VERSION:
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py:35]: The original `self.check_version()` is to ensure customers who use old Azure CLI which installed old azcopy can upgrade their azcopy when we upgrade to latest Azure CLI.

So this should be kept

File: src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py [hunk 2]
```diff
@@ -30,10 +31,26 @@ class AzCopy:
...
+            try:
+                import re
+                from packaging.version import parse as parse_version
+                args = ["azcopy", "--version"]
+                out_bytes = subprocess.check_output(args)
+                out_text = out_bytes.decode('utf-8')
+                version = re.findall(r"azcopy version (.+?)\n", out_text)[0]
+                if parse_version(version) >= parse_version(AZCOPY_VERSION):
+                    self.executable = "azcopy"
+            except FileNotFoundError:
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py:47]: If there's other error, will CLI be crashed? Is it possible that other errors be thrown?

Comment (calvinhzy) [reply src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py]: Using exception to catch all.

File: src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py (hunk 1 above)
Comment (calvinhzy) [reply src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py:35]: Added back.

**Review by evelyn-ys — APPROVED** [review]

**Pull request merged**